plenoptic/simulate/models/frontend.py

```diff
@@ -1,96 +1,407 @@
+        lum = self.luminance(x)
+        lum_normed = linear / (1 + self.luminance_scalar * lum)
 
-    def forward(self, x):
-        x = self._luminance_normalization(x)
-        x = self._contrast_normalization(x)
-        x = self.softplus(x)
+        con = self.contrast(lum_normed.pow(2)).sqrt() + 1E-6  # avoid div by zero
+        con_normed = lum_normed / (1 + self.contrast_scalar * con)
+        y = self.activation(con_normed)
+        return y
 
-        if self._disk is not None and self._disk.shape == x.shape[-2:]:  # uses cached disk_mask if size matches
-            x = self._disk * x
 
-        elif ((self._disk is not None and self._disk.shape != x.shape[-2:])  # create new disk if disk size mismatch
-                or (self._disk is None and self.disk_mask)):  # or if disk does not yet exist
+class OnOff(nn.Module):
+    """Two-channel on-off and off-on center-surround model with local contrast and
+    luminance gain control.
 
-            self._disk = make_disk(x.shape[-1]).to(x.device)
-            x = self._disk * x
+    This model is called OnOff in Berardino et al 2017.
 
-        return x
+    Parameters
+    ----------
+    apply_mask:
+    activation:
+        Activation function following linear and gain control operations.
+
+    Notes
+    -----
+    See `CenterSurround` class for full Parameter docstring.
+
+    Berardino et al., Eigen-Distortions of Hierarchical Representations (2017)
+    http://www.cns.nyu.edu/~lcv/eigendistortions/ModelsIQA.html
+    """
+
+    def __init__(
+        self,
+        kernel_size: Union[int, Tuple[int, int]],
+        width_ratio_limit: float = 4.0,
+        amplitude_ratio: float = 1.25,
+        pad_mode: str = "circular",
+        pretrained=False,
+        activation: Callable[[Tensor], Tensor] = F.softplus,
+        apply_mask: bool = False,
+    ):
+        super().__init__()
+        kernel_size = (31, 31) if pretrained else kernel_size
+
+        self.on = LGG(
+            kernel_size,
+            "on",
+            width_ratio_limit,
+            amplitude_ratio,
+            pad_mode,
+            activation
+        )
+
+        self.off = LGG(
+            kernel_size,
+            "off",
+            width_ratio_limit,
+            amplitude_ratio,
+            pad_mode,
+            activation
+        )
+
+        if pretrained:
+            self.load_state_dict(self._pretrained_state_dict())
+
+        self.apply_mask = apply_mask
+        self._disk = None  # cached disk to apply to image
+
+    def forward(self, x: Tensor) -> Tensor:
+        y = torch.cat((self.on(x), self.off(x)), dim=1)
+
+        if self.apply_mask:
+            im_shape = x.shape[-2:]
+            if self._disk is None or self._disk.shape != im_shape:  # cache new mask
+                self._disk = make_disk(im_shape).to(x.device)
+
+            y = self._disk * y  # apply the mask
+
+        return y
+
+    def display_filters(self, zoom=5.0, **kwargs):
```

Comment:
why does only `OnOff` get a display function? Wouldn't it be useful for all of them?

Reply:
u right. Added to the rest of the models.